pyvips/vimage.py

```diff
@@ -899,6 +1000,230 @@ def toarray(self):
+    #         shape = (self.height, self.width)
+    #     else:
+    #         shape = (self.height, self.width, self.bands)
+    # 
+    #     typestr = FORMAT_TO_TYPESTR[self.format]
+    #     
+    #     interface = {
+    #         'shape': shape,
+    #         'strides': None,
+    #         'descr': [('', typestr)],
+    #         'typestr': typestr,
+    #         'version': 3
+    #     }
+    # 
+    #     return interface
+
+
+    def __array__(self, dtype=None):
+        '''Conversion to a NumPy array.
+        
+        Parameters
+        ----------
+        dtype : str or numpy dtype
+            The dtype to use for the numpy array.  If None, the default dtype of
+            the image is used as defined the global `FORMAT_TO_TYPESTR` dictionary.
+        
+        Returns
+        -------
+        numpy.ndarray
+            The array representation of the image.
+            - Single-band images lose their channel axis.
+            - Single-pixel single-band images are converted to a 0D array.
+            
+
+
+        See https://numpy.org/devdocs/user/basics.dispatch.html for more details.
+
+        This enables a `pyvips.Image` to be used where a numpy array is expected,
+        including in plotting and conversion to other array container types (pytorch, 
+        JAX, dask, etc.), for example.
+
+        `numpy` is a runtime dependency of this function.
+
+        See Also
+        --------
+        `Image.fromarray` for the inverse operation.
+        '''
+        import numpy as np
+
+        arr = (
+            np.frombuffer(self.write_to_memory(), dtype=FORMAT_TO_TYPESTR[self.format])
+            .reshape(self.height, self.width, self.bands)
+        )
+
+        if self.bands == 1:
+            arr = arr.squeeze(axis=-1) # flatten single-band images
+
+            if self.width == 1 and self.height == 1:
+                arr = arr.squeeze(axis=(0, 1)) # flatten single-pixel single-band images
+
+        if dtype is not None:
+            arr = arr.astype(dtype)
+
+        return arr
+
+
+    def numpy(self, dtype=None):
+        '''Convenience function to allow numpy conversion to be at the end of a method chain.
+
+        This mimics the behavior of pytorch: `arr = im.op1().op2().numpy()`
+
+        numpy is a runtime dependency of this function.
+
+        Parameters
+        ----------
+        dtype : str or numpy dtype
+            The dtype to use for the numpy array.  If None, the default dtype of
+            the image is used.
+        
+        Returns
+        -------
+        numpy.ndarray
+            The image as a numpy array.
+
+        See Also
+        --------
+        - `Image.__array__`
+        - `FORMAT_TO_TYPESTR`: Global dictionary mapping libvips format strings to numpy dtype strings.
+        '''
+        return self.__array__(dtype=dtype)
+
+
+    @classmethod
+    def fromarray(cls, obj, interpretation=None):
+        '''Create a new Image from an array-like object.
```

Comment:
It's really unfortunate that pyvips already has `new_from_array` to make an image from a list of lists.

Maybe this method should be `new_from_numpy_array()`?

Reply:
I used the name that PIL had, and my vote would be to keep it a bit more general we we can import things that export the array interface, not just numpy arrays.   I expect that this function will be used a lot (if my grad students are any indication, this and `new_from_file` are the most common), and shorter names are also better. 

The solution we have here is just to define a top-level convenience catch-all called `image` that can take array-likes, Images, lists of Images (which are then `bandjoin`ed to give a single image), strings or Path objects which are treated as filenames, etc.  So then `pyvips.image(something)` just works and it's compact and the underlying name doesn't matter much.

The broader issue is that array seems to mean something special to pyvips (list of lists?) that differs from the usual Python usage.  One possible solution is to modify new_from_array so that it works with either array-like or list-of-lists objects, so that it retains backward compatibility while doing what I think most newcomers expect.  I know that feeding a numpy array to `new_from_array` was the first thing I tried when I discovered pyvips a few years ago,

Reply:
Yes, adding numpy array support to `new_from_array` sounds like a good compromise.